### PR TITLE
QL: complex nested field types mappings fix

### LINKED ISSUE #69175: SQL: double nested fields return null values
For a mapping like the following (one nested field - `address` - inside another nested field - `user`):

```
      "user": {
        "type": "nested",
        "properties": {
          "name" : {
            "type" : "text",
            "fields" : {
              "keyword" : {
                "type" : "keyword"
              }
            }
          },
          "address" : {
            "type" : "nested",
            "properties": {
              "city" : {
                "type" : "keyword"
              },
              "country" : {
                "type" : "keyword"
              }
            }
          }
        }
      }
```

Simply asking for the entire tree returns `null` values.

```
SELECT user.address.city FROM test
```

```
user.address.city
-----------------
null             
null             
null             
null             
```

### PR DESCRIPTION
**Pull request by astefan**
After the changes in https://github.com/elastic/elasticsearch/pull/67432 and https://github.com/elastic/elasticsearch/pull/68802, the way nested fields are extracted from the ES query results changed significantly. This PR fixes the more complex multi-nested mappings fields extraction.

Fixes https://github.com/elastic/elasticsearch/issues/69175.

### CODE REVIEW

**Comment by astefan** [comment]
@elasticmachine update branch

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 1]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
             field = hit.field(fieldName);
+            if (field != null) {
+                value = unwrapFieldsMultiValue(field.getValues());
+            }
+        }
+        return value;
+    }
+
+    /*
+     * For a tree of fields like root.nested1.nested2.leaf where nested1 and nested2 are nested field types,
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:97]: nit: would "a **path** of fields" make more sense? (since I guess the doc isn't parsed/accessed as a tree?)

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 2]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
+     * After this step is done, what remains to be done is just getting the leaf values.
+     */
+    @SuppressWarnings("unchecked")
+    private Object extractNestedField(SearchHit hit) {
+        Object value;
+        DocumentField field;
+        String tempHitname = hitName;
+        List<String> remainingPath = new ArrayList<>();
+        // first, search for the "root" DocumentField under which the remaining path of nested document values is
+        while((field = hit.field(tempHitname)) == null) {
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:110]: nit: spacing after `while`.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 3]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
+            int indexOfDot = tempHitname.lastIndexOf(".");
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:111]: Is this guaranteed to return a non-negative?
Actually, with a mapping similar to the  exemplified one in the test's comment (but only using the first lowered letter of the randomly generated nodes), indexing just `{"h": [{"f":{"b": {"a": "true"}}}]}` and querying with `SELECT  h.f.o FROM test` seems to trigger:
```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 1
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3734)
	at java.base/java.lang.String.substring(String.java:1903)
	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extractNestedField(AbstractFieldHitExtractor.java:113)
	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extract(AbstractFieldHitExtractor.java:86)
	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:129)
	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:32)
	at org.elasticsearch.xpack.sql.execution.search.ResultRowSet.getColumn(ResultRowSet.java:38)
	at org.elasticsearch.xpack.sql.session.AbstractRowSet.column(AbstractRowSet.java:19)
	at org.elasticsearch.xpack.sql.session.RowView.forEachColumn(RowView.java:39)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$createResponse$5(TransportSqlQueryAction.java:132)
	at org.elasticsearch.xpack.sql.session.RowSet.forEachRow(RowSet.java:28)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponse(TransportSqlQueryAction.java:130)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponseWithSchema(TransportSqlQueryAction.java:125)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$operation$0(TransportSqlQueryAction.java:97)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:139)

```

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 4]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
+            remainingPath.add(0, tempHitname.substring(indexOfDot + 1));
+            tempHitname = tempHitname.substring(0, indexOfDot);
         }
-        if (field != null) {
+        // then dig into DocumentField's structure until we reach the field we are interested into
+        if (remainingPath.size() > 0) {
+            List<Object> values = field.getValues();
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:117]: Is it guaranteed to have a non-null `field`? A check, assert or comment as to why this might be the case would help.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 5]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
+            Iterator<String> pathIterator = remainingPath.iterator();
+            while(pathIterator.hasNext()) {
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:119]: nit: spacing.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 6]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
+                String pathElement = pathIterator.next();
+                Map<String, Object> elements = (Map<String, Object>) values.get(0);
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:121]: Here (and when calling `unwrapFieldsMultiValue` below), only a first value is selected. Could the looked for path not be accessed through a subsequent element?

**Review by bpintea — COMMENTED** [review]
A first round of comments.

**Review by matriv — COMMENTED** [review]
Apart from the great randomised test, Imho we should also have some plain tests that target specific cases, as for example those mentioned already by @bpintea

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 3 above)
Comment (bpintea) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:111]: Mentioning it here, but happy to open an issue if better (or drop it, if wrong): selecting `h.f.b.a` will still return nothing.
Reason being that the `inner_hits` response from ES comes with with the entries listed as `h.f_12` (for example), while we're looking for the inner hit `h.f.b` in SearchHitRowSet. Seems like this should be doable to work, but didn't look deeper.
This is not new, this didn't work before either (although no error message is returned).

Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:111]: Nice catch! That's an older problem, for sure. It's the complexity of the `nested` documents definition possibilities that shows that our algorithm isn't covering all of them. I'd be inclined in dealing with this bug in a subsequent PR as it doesn't look straightforward and it may need a bit of rethinking the way we deal with `nested` fields hierarchies.

**Comment by astefan** [comment]
@matriv thank you for the review. I am not sure how useful it is to have a specific test for that scenario. There could be many more scenarios like that one, ones that we didn't even think about, just because the combination of mappings possible is so complex. While we do have specific tests for specific scenarios in `FieldExtractorTestCase`, for `nested` fields the scenarios are not so few also not so simple and I'm reluctant to add a specific case that couldn't be covered by a randomization algorithm.

@bpintea spotted one scenario where the value of the requested field is `null` and I've adjusted the randomized test to account for that. He also noticed another scenario that didn't ever work, where a tree of mappings is nested inside a tree of `nested` fields. Since the fix is not obvious for the moment, I'd suggest dealing with that fix in a separate GH issue where tests should be added as well. If a specific test is proven to be useful in that case vs. a randomized one, I'd let this to be decided then.

File: x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java [hunk 7]
```diff
@@ -1051,6 +1055,167 @@ public void testByteFieldWithIntegerSubfield() throws IOException {
...
+                    // A nested field always has a leaf field, even if not all nested fields in a path will have this value
+                    // indexed. We will index only the "leaf" field of the last nested field in the path, because this is the
+                    // one we will ask back from ES
+                    index.startObject(leafFieldName);
+                    {
+                        index.field("type", "keyword");
+                    }
+                    index.endObject();
+                    // from time to time set a null value instead of an actual value
+                    if (randomBoolean()) {
```
Comment (cbuescher) [inline x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:1173]: maybe use `null` less often than 50% of the times in randomization here, it should be the exception I think. Using ' rarely()` this case would still get picked 1% of the time, should be enough?

File: x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java [hunk 8]
```diff
@@ -1051,6 +1055,167 @@ public void testByteFieldWithIntegerSubfield() throws IOException {
...
+                        randomValue.set(null);
+                        bulkContent.append("\"" + fieldName + "\":{\"" + leafFieldName + "\":null");
+                    } else {
+                        randomValue.set(randomAlphaOfLength(10));
+                        bulkContent.append("\"" + fieldName + "\":{\"" + leafFieldName + "\":\"" + randomValue.get() + "\"");
```
Comment (cbuescher) [inline x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:1178]: I'm not sure I understand this entirely, but to me it seems the test will only add one value for the leaf field thats going to be requested and tested for here? Asking because its very common that there is more than one entry in a nested field, e.g. in a very simple case
```
     "fields": {
        "group" : ["fans"],
        "user": [{
            "first": ["John"],
            "last": ["Smith"],
          },
          {
            "first": ["Alice"],
            "last": ["White"],
          }
        ]
      }
```
where `user` is the nested field, and you want to retrieve `user.first`. I'm not sure this case is tested for, but I expect it to be common. Maybe less common but still relevant would be something like

```
     "fields": {
        "group" : ["fans"],
        "user": [{
            "first": ["John", "Jim", "James"],
            "last": ["Smith"],
          },
          {
            "first": ["Alice"],
            "last": ["White"],
          }
        ]
      }
```
where I would expect probably expect `user.first` to extract a list or array of four values. 
Maybe this is done in `AbstractFieldHitExtractor#unwrapFieldsMultiValue` but I think it would be great to have this explicitely covered by tests as well.

**Review by cbuescher — COMMENTED** [review]
Sorry that I didn't go through all the details of the extraction implementation yet, its a bit different to what we do on the `fields` side of things where we deal with the _source map. I left a question around testing, I tend to agree with @matriv that in addition to the randomized testing it would be great to have a few common cases spelled out explicitely without too much randomness to get a basic coverage of those. But I don't know how well this fits with the general testing philosophy and infra in ql, so I leave this only as a remark.

File: x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java [hunk 9]
```diff
@@ -1051,6 +1055,167 @@ public void testByteFieldWithIntegerSubfield() throws IOException {
...
+                    }
```
Comment (palesz) [inline x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:1179]:
```suggestion
                    randomValue.set(randomBoolean() ? null : randomAlphaOfLength(10));
                    bulkContent.append("\"" + fieldName + "\":{\"" + leafFieldName + "\":\"" + randomValue.get() + "\"");
```

**Review by palesz — COMMENTED** [review]
Couple of minor comments on top of the already existing reviews.

File: x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java (hunk 9 above)
Comment (astefan) [reply x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:1179]: There is tiny difference between my approach and yours: when setting a `null` value in json the json looks like this `"field_name": null`. Your suggestion leads to a json looking like `"field_name": "null"` which is the string `null`, not a `null` value. I know the `if else` approach is a more verbose one , but otherwise the string being built inside `bulkContent` would have looked messy and hard to follow.

Comment (palesz) [reply x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:1179]: Yep, missed that. Fair point.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 10]
```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
...
+                 * if this path is not found it means we hit another nested document (inner_root_1.inner_root_2.nested_field_2)
+                 * something like this
+                 * "root_field_1.root_field_2.nested_field_1" : [
+                 *     {
+                 *       "inner_root_1.inner_root_2.nested_field_2" : [
+                 *         {
+                 *           "leaf_field" : [
+                 *             "abc2"
+                 *           ]
+                 * So, start re-building the path until the right one is found, ie inner_root_1.inner_root_2......
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:136]: The SearchHit structure for a `field` (as in asking for `SearchHit.field(fieldName)`) looks like a series of `List`s and `Map`s: List of Maps where maps have as key a string (the name of a field/subfield) and the value can be another set of Lists of Maps.

If we have to look for a field called `a.b.c.d.e` (this is represented as AbstractFieldHitExtractor.fieldName) and this field sits under a `nested` data structure (and implicitly has `AbstractFieldHitExtractor.hitName` != null) with `hitName` = `a.b.c`, the algorithm searches for `a.b.c` (SearchHit.field("a.b.c")), if this is null then we search for `SearchHit.field("a.b")` and if still not found we search for `SearchHit.field("a")`. When we find a DocumentField, we take its `values` and we start searching in its List of Maps for the rest of the field.
If we found a DocumentField for `a` then we still need to search for the rest of the field name path, which now is `b.c.d.e`. So, we look in `values` for `b`, then for `b.c` and so on until we find an element. If we find `b.c` then the rest of the path becomes `d.e` and we repeat the process.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 11]
```diff
@@ -81,12 +83,65 @@ public Object extract(SearchHit hit) {
...
+                 */
+                while (values == null) {
+                    pathElement += "." + pathIterator.next();
+                    values = (List<Object>) elements.get(pathElement);
+                }
+            }
+            value = unwrapFieldsMultiValue(((Map<String, Object>) values.get(0)).get(fieldName.substring(hitName.length() + 1)));
+        } else {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:144]: Since the outer code also calls unwrapFieldMultiValue, how about returning its argument instead:
```
value = ((Map<String, Object>) values.get(0)).get(fieldName.substring(hitName.length() + 1));

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 12]
```diff
@@ -81,12 +83,65 @@ public Object extract(SearchHit hit) {
...
         DocumentField field = null;
         if (hitName != null) {
-            // a nested field value is grouped under the nested parent name (ie dep.dep_name lives under "dep":[{dep_name:value}])
-            field = hit.field(hitName);
+            value = extractNestedField(hit);
         } else {
             field = hit.field(fieldName);
+            if (field != null) {
+                value = unwrapFieldsMultiValue(field.getValues());
+            }
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:91]: This should be common for both branches (needs changes to `extractNestedField`) to make it more consistent. Something like:
```
field = hitName != null  ? extractNestedField(hit) : hit.field(fieldName);
if (field != null) {
    value = unwrapFieldsMultiValue(fields.getValue());
}
```

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:111]: > Is this guaranteed to return a non-negative?
> Actually, with a mapping similar to the exemplified one in the test's comment (but only using the first lowered letter of the randomly generated nodes), indexing just `{"h": [{"f":{"b": {"a": "true"}}}]}` and querying with `SELECT h.f.o FROM test` seems to trigger:
> 
> ```
> Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 1
> 	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3734)
> 	at java.base/java.lang.String.substring(String.java:1903)
> 	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extractNestedField(AbstractFieldHitExtractor.java:113)
> 	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extract(AbstractFieldHitExtractor.java:86)
> 	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:129)
> 	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:32)
> 	at org.elasticsearch.xpack.sql.execution.search.ResultRowSet.getColumn(ResultRowSet.java:38)
> 	at org.elasticsearch.xpack.sql.session.AbstractRowSet.column(AbstractRowSet.java:19)
> 	at org.elasticsearch.xpack.sql.session.RowView.forEachColumn(RowView.java:39)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$createResponse$5(TransportSqlQueryAction.java:132)
> 	at org.elasticsearch.xpack.sql.session.RowSet.forEachRow(RowSet.java:28)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponse(TransportSqlQueryAction.java:130)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponseWithSchema(TransportSqlQueryAction.java:125)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$operation$0(TransportSqlQueryAction.java:97)
> 	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:139)
> ```

We should prevent the error from occurring - looking at the stacktrace this seems to have been addressed yet I'm mentioning to double check this is indeed the case.

**Review by costin — APPROVED** [review]
LGTM - it's unfortunate the reading of nested docs is messy and there are no built-in utilities that we can leverage and instead need to reconstruct the hierarchy ourselves through strings which are error prone.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 12 above)
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:91]: `extractNestedField` cannot consistently return a `DocumentField` instance so that we can call allways `field.getValues()` in `extract(SearchHit)` method. The data structure inside a `DocumentField` is more complex in the case of nested fields and it requires further exploration before obtaining an equivalent of `field.getValues()`  (see https://github.com/elastic/elasticsearch/pull/69368/files#diff-099b1feb3ed943ecc359be919a1dd67423938b9b8252e43c9ddd0a72027c1c87R120-R143).

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:111]: @bpintea [this change](https://github.com/elastic/elasticsearch/pull/69368/files#diff-85f0eb71298942313786f64c1d08628baec44a09fcd1c70fa1ee46a4dfad335aR341) should fix the `h.f.b.a` returning nothing example you mentioned. Test for this exact case [here](https://github.com/elastic/elasticsearch/pull/69368/files#diff-990981c3256841bd9e0cc66efff91f49397139b1ae9cf548a8a6cae8c71fe5fdR1139).

**Review by bpintea — APPROVED** [review]
LGTM

**Review by matriv — APPROVED** [review]
LGTM, Thank you!
